File: lab7/q2.cpp
#include<iostream>
#include<thread>
#include<mutex>
#include<stdlib.h>
using namespace std;
int amount = 600;
std::mutex mtx;
class father{
	public:
		void operator()(int x){
			while(1){
				deposit();	
			}
		}
		void deposit(){
			mtx.lock();
			if(amount<2000){
				int temp = rand()%200 + 1;
				amount += temp;
				cout<<"Father added "<<temp<<" Amount now = "<<amount<<endl;
			}	
			mtx.unlock();
		}
};
class son{
	public:
		void operator()(int x){
			while(1){
				withdraw();
			}
		}
		void withdraw(){
			mtx.lock();
			if(amount>500){
				int temp = rand()%150 + 1;
				amount -= temp;
				cout<<"Son withdrew "<<temp<<" Amount now = "<<amount<<endl;
			}
			mtx.unlock();
		}
};

int main(){
	thread th1(father(),1);
	thread th2(son(),1);
	th1.join();
	th2.join();
	return 0;
}

File: lab7/q2.java
import java.util.Random;
class Father extends Thread
{
	Random rand = new Random();
	public void run(){
		while(true){
			deposit();
		}
	}
	public synchronized void deposit(){
		if(q2.balance <= 2000){
			int r = rand.nextInt(200);
			q2.balance += r;
			System.out.println("Father deposited Rs " + r + ". New Balance is Rs " + q2.balance + ".");
		}
	}
}
class Son extends Thread{
	Random rand = new Random();
	public void run(){
		while(true){
			withdraw();	
		}
	}
	public synchronized void withdraw(){
		
		if(q2.balance >= 500){
			int r = rand.nextInt(150);
			q2.balance -= r;
			System.out.println("Son withdrawed Rs " + r + ". New Balance is Rs " + q2.balance + ".");
		}
	}
}
class q2{
	public static int balance = 600;
	public static void main(String[] args){
		Father f = new Father();
		Son s = new Son();
		f.start();
		s.start();
	}
}
